src/components/shared/MarkerTooltipContents.js

```diff
@@ -37,19 +39,80 @@ function _markerDetail<T: NotVoidOrNull>(
   fn: T => string = String
 ): React.Node {
   return [
-    <div className="tooltipLabel" key="{key}">
+    <div className="tooltipLabel" key={key}>
       {label}:
     </div>,
     fn(value),
   ];
 }
 
+type PhaseTimeTuple = { name: string, time: Microseconds };
+
+function _markerDetailPhase(p: PhaseTimeTuple): React.Node {
+  return _markerDetail(
+    'gcphase' + p.name,
+    'Phase ' + p.name,
+    p.time,
+    t => formatNumber(t / 1000) + 'ms'
+  );
+}
+
+function _makePhaseTimesArray(
+  phases: PhaseTimes<Microseconds>
+): Array<PhaseTimeTuple> {
+  const array = [];
+  for (const phase in phases) {
+    if (phase !== 'Total') {
+      array.push({ name: phase, time: phases[phase] });
+    }
+  }
+  return array;
+}
+
+function _leafPhase(phases: Array<PhaseTimeTuple>, name: string): boolean {
+  for (let i = 0; i < phases.length; i++) {
+    const pt = phases[i];
+    if (pt.name === name) {
+      // Don't consider itself.
+      continue;
+    }
+    if (pt.name.startsWith(name)) {
+      // Name is not a leaf
+      return false;
+    }
+  }
+  // name is a leaf.
+  return true;
```

Comment:
I also wondering why we check if the name _doesn't start_ with `name` instead of checking if it starts with  `name` (using `endsWith`. I wonder if we shouldn't use:
```javascript
pt.name.endsWith(`.${name}`)
```

Reply:
That first bit is true.

But I don't think that second idea will work, name, now phaseName, is always the full name.